File: packages/jvm/src/main/java/io/github/qorechain/rdk/bridge/Withdrawal.java
package io.github.qorechain.rdk.bridge;

import io.github.qorechain.rdk.bridge.Merkle.MerkleOptions;
import io.github.qorechain.rdk.bridge.Merkle.MerkleProof;
import io.github.qorechain.rdk.tx.Messages.ExecuteWithdrawalInput;
import java.util.List;

/**
 * Withdrawal-proof assembly for {@code MsgExecuteWithdrawal}.
 *
 * <p>A withdrawal is executed by proving its leaf is committed in a finalized batch's
 * {@code withdrawals_root}. This helper turns the batch's withdrawal leaves into the sibling-hash
 * proof the message carries. The leaf encoding and hashing must match the network's construction —
 * see {@link MerkleOptions}.
 */
public final class Withdrawal {
    private Withdrawal() {}

    /** The proof material for a single withdrawal. */
    public static final class WithdrawalProof {
        /** Sibling hashes from the leaf to {@code withdrawals_root}, for the {@code proof} field. */
        public final List<byte[]> proof;
        /** The computed {@code withdrawals_root} (compare against the batch's). */
        public final byte[] withdrawalsRoot;
        /** The withdrawal's index within the batch. */
        public final int withdrawalIndex;

        public WithdrawalProof(List<byte[]> proof, byte[] withdrawalsRoot, int withdrawalIndex) {
            this.proof = proof;
            this.withdrawalsRoot = withdrawalsRoot;
            this.withdrawalIndex = withdrawalIndex;
        }
    }

    /**
     * Assemble the Merkle proof for the withdrawal at {@code withdrawalIndex} from the full list of
     * the batch's withdrawal leaves.
     */
    public static WithdrawalProof assembleWithdrawalProof(
            List<byte[]> leaves, int withdrawalIndex, MerkleOptions options) {
        MerkleProof proof = Merkle.binaryMerkleProof(leaves, withdrawalIndex, options);
        return new WithdrawalProof(proof.siblings, proof.root, withdrawalIndex);
    }

    /**
     * Combine a withdrawal's recipient/amount details with an assembled proof into the
     * {@link ExecuteWithdrawalInput} for the tx client.
     */
    public static ExecuteWithdrawalInput buildExecuteWithdrawalInput(
            String submitter,
            String rollupId,
            long batchIndex,
            String recipient,
            String denom,
            long amount,
            WithdrawalProof withdrawal) {
        ExecuteWithdrawalInput in = new ExecuteWithdrawalInput();
        in.submitter = submitter;
        in.rollupId = rollupId;
        in.batchIndex = batchIndex;
        in.withdrawalIndex = withdrawal.withdrawalIndex;
        in.recipient = recipient;
        in.denom = denom;
        in.amount = amount;
        in.proof = withdrawal.proof;
        return in;
    }
}

File: packages/py/src/qorrdk/bridge/withdrawal.py
"""Withdrawal-proof assembly for ``MsgExecuteWithdrawal``.

A withdrawal is executed by proving its leaf is committed in a finalized batch's
``withdrawals_root``. This helper turns the batch's withdrawal leaves into the
sibling-hash proof the message carries. The leaf encoding and hashing must match
the network's construction -- see :class:`MerkleOptions`.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Union

from ..tx.messages import ExecuteWithdrawalInput
from .merkle import MerkleOptions, binary_merkle_proof


@dataclass
class WithdrawalProof:
    """The proof material for a single withdrawal."""

    #: Sibling hashes from the leaf to ``withdrawals_root``, for the ``proof`` field.
    proof: list[bytes]
    #: The computed ``withdrawals_root`` (compare against the batch's).
    withdrawals_root: bytes
    #: The withdrawal's index within the batch.
    withdrawal_index: int


def assemble_withdrawal_proof(
    leaves: list[bytes],
    withdrawal_index: int,
    options: Optional[MerkleOptions] = None,
) -> WithdrawalProof:
    """Assemble the Merkle proof for the withdrawal at ``withdrawal_index``."""
    result = binary_merkle_proof(leaves, withdrawal_index, options)
    return WithdrawalProof(
        proof=result.siblings,
        withdrawals_root=result.root,
        withdrawal_index=withdrawal_index,
    )


def build_execute_withdrawal_input(
    *,
    submitter: str,
    rollup_id: str,
    batch_index: Union[int, str],
    recipient: str,
    denom: str,
    amount: Union[int, str],
    withdrawal: WithdrawalProof,
) -> ExecuteWithdrawalInput:
    """Combine a withdrawal's details with an assembled proof into the tx input."""
    return ExecuteWithdrawalInput(
        submitter=submitter,
        rollup_id=rollup_id,
        batch_index=batch_index,
        withdrawal_index=withdrawal.withdrawal_index,
        recipient=recipient,
        denom=denom,
        amount=amount,
        proof=list(withdrawal.proof),
    )


__all__ = [
    "WithdrawalProof",
    "assemble_withdrawal_proof",
    "build_execute_withdrawal_input",
]
